File: src/main/java/com/zhihao/admin/module/system/service/UserService.java
package com.zhihao.admin.module.system.service;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.zhihao.admin.common.api.PageResult;
import com.zhihao.admin.common.exception.BusinessException;
import com.zhihao.admin.module.system.dto.PasswordResetRequest;
import com.zhihao.admin.module.system.dto.UserCreateRequest;
import com.zhihao.admin.module.system.dto.UserDetailResponse;
import com.zhihao.admin.module.system.dto.UserQuery;
import com.zhihao.admin.module.system.dto.UserUpdateRequest;
import com.zhihao.admin.module.system.entity.SysUser;
import com.zhihao.admin.module.system.entity.SysUserRole;
import com.zhihao.admin.module.system.mapper.SysUserMapper;
import com.zhihao.admin.module.system.mapper.SysUserRoleMapper;
import java.util.List;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

@Service
@RequiredArgsConstructor
public class UserService {

    private final SysUserMapper userMapper;
    private final SysUserRoleMapper userRoleMapper;
    private final PasswordEncoder passwordEncoder;

    public PageResult<UserDetailResponse> page(UserQuery query) {
        LambdaQueryWrapper<SysUser> wrapper = new LambdaQueryWrapper<SysUser>()
                .like(StringUtils.hasText(query.getUsername()), SysUser::getUsername, query.getUsername())
                .like(StringUtils.hasText(query.getPhone()), SysUser::getPhone, query.getPhone())
                .eq(query.getStatus() != null, SysUser::getStatus, query.getStatus())
                .orderByDesc(SysUser::getCreateTime);
        Page<SysUser> page = userMapper.selectPage(Page.of(query.getPageNum(), query.getPageSize()), wrapper);
        Page<UserDetailResponse> mapped = new Page<>(page.getCurrent(), page.getSize(), page.getTotal());
        mapped.setRecords(page.getRecords().stream()
                .map(u -> UserDetailResponse.from(u, null))
                .toList());
        return PageResult.from(mapped);
    }

    public UserDetailResponse getDetail(Long id) {
        SysUser user = requireUser(id);
        List<Long> roleIds = userRoleMapper.selectList(
                new LambdaQueryWrapper<SysUserRole>().eq(SysUserRole::getUserId, id)
        ).stream().map(SysUserRole::getRoleId).toList();
        return UserDetailResponse.from(user, roleIds);
    }

    @Transactional(rollbackFor = Exception.class)
    public Long create(UserCreateRequest request) {
        ensureUsernameAvailable(request.username(), null);
        SysUser user = new SysUser();
        user.setDeptId(request.deptId());
        user.setUsername(request.username());
        user.setNickname(request.nickname());
        user.setPassword(passwordEncoder.encode(request.password()));
        user.setEmail(request.email());
        user.setPhone(request.phone());
        user.setStatus(request.status());
        user.setRemark(request.remark());
        userMapper.insert(user);
        replaceRoles(user.getId(), request.roleIds());
        return user.getId();
    }

    @Transactional(rollbackFor = Exception.class)
    public void update(Long id, UserUpdateRequest request) {
        SysUser user = requireUser(id);
        user.setDeptId(request.deptId());
        user.setNickname(request.nickname());
        user.setEmail(request.email());
        user.setPhone(request.phone());
        user.setStatus(request.status());
        user.setRemark(request.remark());
        userMapper.updateById(user);
        replaceRoles(id, request.roleIds());
    }

    @Transactional(rollbackFor = Exception.class)
    public void delete(Long id) {
        SysUser user = requireUser(id);
        user.setUsername(user.getUsername() + ":" + user.getId());
        userMapper.updateById(user);
        userRoleMapper.delete(new LambdaQueryWrapper<SysUserRole>().eq(SysUserRole::getUserId, id));
        userMapper.deleteById(id);
    }

    public void resetPassword(Long id, PasswordResetRequest request) {
        SysUser user = requireUser(id);
        user.setPassword(passwordEncoder.encode(request.password()));
        userMapper.updateById(user);
    }

    private SysUser requireUser(Long id) {
        SysUser user = userMapper.selectById(id);
        if (user == null) {
            throw new BusinessException("user not found");
        }
        return user;
    }

    private void ensureUsernameAvailable(String username, Long ignoreId) {
        SysUser user = userMapper.selectOne(new LambdaQueryWrapper<SysUser>()
                .eq(SysUser::getUsername, username)
                .ne(ignoreId != null, SysUser::getId, ignoreId)
                .last("limit 1"));
        if (user != null) {
            throw new BusinessException("username already exists");
        }
    }

    private void replaceRoles(Long userId, List<Long> roleIds) {
        userRoleMapper.delete(new LambdaQueryWrapper<SysUserRole>().eq(SysUserRole::getUserId, userId));
        if (roleIds == null || roleIds.isEmpty()) {
            return;
        }
        List<SysUserRole> userRoles = roleIds.stream()
                .distinct()
                .map(roleId -> {
                    SysUserRole ur = new SysUserRole();
                    ur.setUserId(userId);
                    ur.setRoleId(roleId);
                    return ur;
                })
                .collect(Collectors.toList());
        userRoleMapper.insertBatch(userRoles);
    }
}

File: server/src/modules/user/user.service.ts
import bcrypt from 'bcryptjs'
import { pageSkip, toPageResult, type PageResult } from '../../common/page'
import { BusinessException } from '../../common/errors'
import { prisma } from '../../lib/prisma'
import type {
  PasswordResetRequest,
  UserCreateRequest,
  UserDetailResponse,
  UserQuery,
  UserUpdateRequest,
} from './user.schema'

function toUserResponse(
  user: {
    id: bigint
    deptId: bigint | null
    username: string
    nickname: string
    email: string | null
    phone: string | null
    status: number
    remark: string | null
  },
  roleIds: number[] | null,
): UserDetailResponse {
  return {
    id: Number(user.id),
    deptId: user.deptId !== null ? Number(user.deptId) : null,
    username: user.username,
    nickname: user.nickname,
    email: user.email ?? '',
    phone: user.phone ?? '',
    status: user.status,
    remark: user.remark ?? '',
    roleIds,
  }
}

export async function page(query: UserQuery): Promise<PageResult<UserDetailResponse>> {
  const where = {
    deleted: 0,
    ...(query.username ? { username: { contains: query.username } } : {}),
    ...(query.phone ? { phone: { contains: query.phone } } : {}),
    ...(query.status !== undefined ? { status: query.status } : {}),
  }

  const [total, records] = await Promise.all([
    prisma.sysUser.count({ where }),
    prisma.sysUser.findMany({
      where,
      orderBy: { createTime: 'desc' },
      skip: pageSkip(query.pageNum, query.pageSize),
      take: query.pageSize,
    }),
  ])

  return toPageResult(
    total,
    query.pageNum,
    query.pageSize,
    records.map((user) => toUserResponse(user, null)),
  )
}

export async function getDetail(id: number): Promise<UserDetailResponse> {
  const user = await requireUser(id)
  const userRoles = await prisma.sysUserRole.findMany({ where: { userId: BigInt(id) } })
  const roleIds = userRoles.map((item) => Number(item.roleId))
  return toUserResponse(user, roleIds)
}

export async function create(request: UserCreateRequest, operatorId: number | null): Promise<number> {
  await ensureUsernameAvailable(request.username, null)
  const now = new Date()
  const user = await prisma.sysUser.create({
    data: {
      deptId: request.deptId !== undefined ? BigInt(request.deptId) : null,
      username: request.username,
      nickname: request.nickname,
      password: bcrypt.hashSync(request.password, 10),
      email: request.email || null,
      phone: request.phone || null,
      status: request.status,
      remark: request.remark,
      createTime: now,
      updateTime: now,
      createBy: operatorId !== null ? BigInt(operatorId) : null,
      updateBy: operatorId !== null ? BigInt(operatorId) : null,
      deleted: 0,
    },
  })
  await replaceRoles(Number(user.id), request.roleIds)
  return Number(user.id)
}

export async function update(id: number, request: UserUpdateRequest, operatorId: number | null): Promise<void> {
  await requireUser(id)
  await prisma.sysUser.update({
    where: { id: BigInt(id) },
    data: {
      deptId: request.deptId !== undefined ? BigInt(request.deptId) : null,
      nickname: request.nickname,
      email: request.email || null,
      phone: request.phone || null,
      status: request.status,
      remark: request.remark,
      updateTime: new Date(),
      updateBy: operatorId !== null ? BigInt(operatorId) : null,
    },
  })
  await replaceRoles(id, request.roleIds)
}

export async function remove(id: number): Promise<void> {
  const user = await requireUser(id)
  await prisma.sysUser.update({
    where: { id: BigInt(id) },
    data: { username: `${user.username}:${user.id}`, updateTime: new Date() },
  })
  await prisma.sysUserRole.deleteMany({ where: { userId: BigInt(id) } })
  await prisma.sysUser.update({
    where: { id: BigInt(id) },
    data: { deleted: 1, updateTime: new Date() },
  })
}

export async function resetPassword(id: number, request: PasswordResetRequest): Promise<void> {
  await requireUser(id)
  await prisma.sysUser.update({
    where: { id: BigInt(id) },
    data: {
      password: bcrypt.hashSync(request.password, 10),
      updateTime: new Date(),
    },
  })
}

async function replaceRoles(userId: number, roleIds: number[] | undefined) {
  await prisma.sysUserRole.deleteMany({ where: { userId: BigInt(userId) } })
  if (!roleIds || roleIds.length === 0) {
    return
  }
  const uniqueRoleIds = [...new Set(roleIds)]
  await prisma.sysUserRole.createMany({
    data: uniqueRoleIds.map((roleId) => ({
      userId: BigInt(userId),
      roleId: BigInt(roleId),
    })),
  })
}

async function requireUser(id: number) {
  const user = await prisma.sysUser.findFirst({
    where: { id: BigInt(id), deleted: 0 },
  })
  if (!user) {
    throw new BusinessException('user not found')
  }
  return user
}

async function ensureUsernameAvailable(username: string, ignoreId: number | null) {
  const user = await prisma.sysUser.findFirst({
    where: {
      username,
      deleted: 0,
      ...(ignoreId !== null ? { id: { not: BigInt(ignoreId) } } : {}),
    },
  })
  if (user) {
    throw new BusinessException('username already exists')
  }
}
